src/python/WMCore/MicroService/MSCore/MSAuth.py

```diff
@@ -0,0 +1,115 @@
+"""
+
+# system modules
+import os
+import json
+
+# third party modules
+import cherrypy
+
+# WMCore modules
+from WMCore.MicroService.Tools.Common import getMSLogger
+from WMCore.REST.Auth import user_info_from_headers
+
+
+def readAuthzRules(entry):
+    """
+    Return authz rules from given entry
+    :param entry: either file name or list of authz rules
+    :return: list of authz rules
+    """
+    rules = []
+    if entry is None:
+        return rules
+    # of provided is a list data-type it may contaisn our rules
+    if isinstance(entry, list) and len(entry) > 0:
+        if isinstance(entry[0], dict):
+            return entry
+    # check if entry is a file name
+    if os.path.exists(entry):
+        with open(entry, 'r', encoding="utf-8") as fstream:
+            rules = json.load(fstream)
+            return rules
+    msg = f"Unable to read {entry} as authz rules"
+    raise Exception(msg)
+
+
+class MSAuth():
+    """
+    This class provides auth/authz functionality for micro-services
+    """
+
+    def __init__(self, msConfig, **kwargs):
+        """
+        Provides a basic setup for all the microservices
+
+        :param config: MS service configuration
+        :param kwargs: optional parameters
+        """
+        self.logger = getMSLogger(getattr(msConfig, 'verbose', False), kwargs.get("logger"))
+        self.msConfig = msConfig
+        self.logger.info("Configuration including default values:\n%s", self.msConfig)
+
+        # load auth/authz configuration
+        self.authzRules = readAuthzRules(msConfig.get('authz_rules', None))
+        self.authzKey = msConfig.get('authz_key', None)
+
+    def authorizeApiAccess(self, service, action, method=None):
+        """
+        Check auth role.
+        :return: boolean
+        """
+        # get user information
+        user = user_info_from_headers(key=self.authzKey)
+
+        # CMS AUTH headers
+        # cms-auth-status, cms-authn, cms-authz, cms-authn-hmac, cms-authn-name, cms-authn-dn
+        match = False
+        for entry in self.authzRules:
+            # check that user method is allowed
+            if method and method not in entry['method']:
+                continue
+            userMethod = user.get('method', None)
+            if userMethod and userMethod not in entry['method']:
+                continue
+
+            # check if entry service and action matches
+            if entry['service'] == service and entry['action'] == action:
+                # if we have match for service and action we'll check used roles
+                # 'roles': {'fake': {'site': set(['fake']), 'group': set(['fake'])}}
+                for role, obj in user['roles'].items():
+                    if role != entry['role']:
+                        continue
+                    for group in obj['group']:
+                        if group == entry['group']:
+                            match = True
+
+        if not match:
```

Comment:
and here simply remove the match and raise an exception if it gets to this point.

Reply:
done